src/js/fs.js

```diff
@@ -471,3 +484,101 @@ function checkArgString(value, name) {
 function checkArgFunction(value, name) {
   return checkArgType(value, name, util.isFunction);
 }
+
+
+function ReadStream(path, options) {
+  if (!(this instanceof ReadStream)) {
+    return new ReadStream(path, options);
+  }
+
+  var defaultOptions = {
+    flags: 'r',
+    encoding: null,
+    fd: null,
+    mode: 0666,
+    autoClose: true,
+    bufferSize: 65536,
+  };
+
+  options = options || defaultOptions;
+
+  Readable.call(this, {defaultEncoding: options.encoding});
+
+  this._buffer = new Buffer(options.bufferSize);
+
+  var self = this;
+  fs.open(path, options.flags, options.mode, function(err, _fd) {
+    self._fd = _fd;
+
+    fs.read(self._fd, self._buffer, 0, self._buffer.length, -1, function(err, bytes_read, buffer) {
+      if (err) {
+        throw new Error("Error reading file");
+      }
+
+      if (bytes_read !== 0) {
+        self.push(self._buffer.slice(0, bytes_read));
```

Comment:
Why slice is needed?

Reply:
It is needed because `self._buffer` may be bigger than the number of bytes we read, so there will be some zeros at the end of the buffer, which should not be `push`ed. This caused a problem when piping a readable filestream into a writable filestream, because the whole buffer will be written to the output file, so the file will contain all those zeros at the end. So for example:
```js
var fs = require('fs');

// let's say in.txt contains "hello world"
var rStream = fs.createReadStream('in.txt');
var wStream = fs.createWriteStream('out.txt');
wStream._readyToWrite();

rStream.pipe(wStream);
// out.txt will contain "hello world" and a lot of zeros at the end, depending on the buffer size of rStream
```